Adding empty state component when there are no filter matches

amarie401 opened the pull request:
An empty state message when the filters result in no matches while also allowing users to clear the applied filters through the url [by adding an optional urlAction property to the angular-pf emptystate component](https://github.com/patternfly/angular-patternfly/pull/648)

NOTE: A-PF will need to be updated to introduce this new config property in order for the Clear Filters link to work.

https://github.com/patternfly/angular-patternfly/pull/650

This would close #461 

<img width="1285" alt="screen shot 2017-10-19 at 1 28 54 pm" src="https://user-images.githubusercontent.com/20052391/31784936-8282cc14-b4d1-11e7-93c8-0ac8210d77ea.png">



@spadgett @dtaylor113 @beanh66

beanh66 commented:
@amarie401 The styling looks great! After looking through more OpenShift and PatternFly empty state examples, wondering if we can we make a slight adjustment to the copy to simplify? Suggestion below:

No results match. (title)
The active filters are hiding all catalog items. (description text) 
Clear Filters (link)

spadgett commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -63,6 +63,17 @@ export class ServicesViewController implements angular.IController {
+
+    this.ctrl.emptyFilterConfig = {
+      icon: 'pficon-info',
+      title: 'No results match the filter criteria',
```
Per @beanh66 `'No results match.'`

spadgett commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -63,6 +63,17 @@ export class ServicesViewController implements angular.IController {
+      title: 'No results match the filter criteria',
+      info: 'The active filters are currently hiding all catalog items.',
+      helpLink: {
+        label: 'To view all items',
```
I would leave `label` blank so you just see "Clear Filters"

spadgett commented:
@amarie401 Sorry, just realized you probably haven't updated the PR yet since @beanh66's comment!

amarie401 commented:
I have made requested changes @spadgett @dtaylor113 @beanh66

amarie401 commented:
I've tried a few different combinations and they seem to pass in karma so not sure why the test is failing in Travis..

I've tried:

```1) var filterLink = jQuery(element).find('pf-empty-state');
      filterLink.find('.blank-state-pf-helpLink').click();
      expect(jQuery(element).find('.active-filter.label.label-info').length).toBe(0);

      2) var filterLink = jQuery(element).find('.blank-slate-pf');
      filterLink.find('.blank-state-pf-helpLink').click();
      expect(jQuery(element).find('.active-filter').length).toBe(0);

      3) componentTest.eventFire(element.querySelector('.blank-state-pf-helpLink'), 'click');
      expect(jQuery(element).find('.active-filter.label.label-info').length).toBe(0);

jeff-phillips-18 commented:
Try  adding `componentTest.scope.$digest();` after the click.

amarie401 commented:
I've added that to a few of the combinations I've tried but still failing :/

dtaylor113 commented:
Hi, looking into this...

amarie401 commented:
After looking into it, it seems like firefox is causing the unit test issues that I am experiencing. When changing the test config to use firefox it then fails. I noticed this was brought up in issue #433 . What would be the best approach?

@dtaylor113 @spadgett @jeff-phillips-18

dtaylor113 commented:
Hi @amarie401,
I'm using FireFox v49 and the unit tests pass when run locally.  I suspect it is an event timing issue.  When the 'Clear Filters' link is clicked it asynchronously broadcasts an event, which maybe happening after the synchronous unit test has run.

In the failing unit test, you have two ".toBe(0)" expectations.  I'd add a log message to each one.  For example: 
`expect(jQuery(element).find('pf-empty-state').length).toEqual(0, 'empty state not shown);` 
...and...
`expect(jQuery(element).find('.active-filter.label.label-info').length).toBe(0, 'no filter tags');`.
...just to verify which expect 0 is actually failing.

Next, you can have the unit test spit out the HTML it is seeing by adding:
`console.log(jQuery(element)[0].innerHTML);`
...right before the last expect(..) in the unit test.

I was going to suggest that it maybe better to test that the `clearAppliedFilters` method is being called via:
```
var ctrl = componentTest.isoScope.$ctrl;
spyOn(ctrl, 'clearAppliedFilters');
//Click clear filter link
...
expect(ctrl.clearAppliedFilters).toHaveBeenCalled();
```
However, I get a failure that clearAppliedFilters wasn't called.   In general, if you can't get this unit test to run, I'd just remove it for this PR and perhaps follow up with another PR which adds in this unit test.

rhamilto commented on src/components/services-view/services-view.html:
```diff
@@ -100,6 +100,9 @@
                           filter-on-keyword="$ctrl.keywordFilterValue"
                           apply-filters="$ctrl.applyFilters($event)">
           </catalog-filter>
+          <div class="col-xs-12" ng-if="$ctrl.filterConfig.appliedFilters.length > 0 && $ctrl.filterConfig.resultsCount === 0">
```
class="col-xs-12" should not be needed.

amarie401 replied:
I added that because without it then the pfEmptyState doesn't span the full width. Any alternatives that you suggest?

amarie401 replied:
<img width="903" alt="screen shot 2017-10-18 at 3 43 08 pm" src="https://user-images.githubusercontent.com/20052391/31739179-1b0adf7e-b41b-11e7-829e-d4955ba0edfb.png">

rhamilto replied:
Ah.  I'd be inclined to write a rule for this case.  col-xs-12 works to get the appearance you're looking for, but isn't really intended for this application.

amarie401 replied:
Could you clarify what you mean by write a rule for this case?

amarie401 replied:
I see what you're saying. I'll go back and fix.

rhamilto replied:
At https://github.com/openshift/origin-web-catalog/blob/master/src/styles/services-view.less#L13, add:

```
.pf-empty-state {
  margin: 0 (@grid-gutter-width / 2);
  width: 100%;
}
```

and then change your [markup class](https://github.com/amarie401/origin-web-catalog/blob/02c97bc365b6d9ad77ce76c16205328a0300eded/src/components/services-view/services-view.html#L103) from `col-xs-12` to `pf-empty-state`.

You also need to include &lt;pf-empty-state&gt; for [subcategories](https://github.com/amarie401/origin-web-catalog/blob/02c97bc365b6d9ad77ce76c16205328a0300eded/src/components/services-view/services-view.html#L78)?

![screen shot 2017-10-18 at 4 25 13 pm](https://user-images.githubusercontent.com/895728/31741087-2af9331c-b421-11e7-8e2b-0231a1eed507.PNG)

rhamilto commented:
The "Clear Filters" link inside pf-empty-state doesn't appear to work.

amarie401 commented:
@rhamilto I made your suggested changes to the less. Also I'm not sure why the Clear Filters link isn't working for you. It is working on my local. Have you pulled down the emptystate-filter branch?

rhamilto commented:
> Have you pulled down the emptystate-filter branch?

Yep.  And I updated the clearAppliedFilters function to console log each time it is called, which is only when I navigate categories and subcategories.  :(

rhamilto commented:
https://github.com/patternfly/angular-patternfly/commit/23e37790 is why yours works on mine doesn't.  Please note the dependency in your PR.  This PR should not merge without the dependency update!

amarie401 commented:
@rhamilto ahhh yes. In this PR I had linked to the a-pf PR where I introduced the urlAction property to the empty-state config to be used for this issue. Will make sure to note that more clearly in this PR. Thanks!

jeff-phillips-18 commented:
@amarie401 Can you update the version of angular-patternfly to 4.11.5 as part of this PR?

amarie401 commented:
@jeff-phillips-18 I can 👍

jeff-phillips-18 commented:
https://github.com/openshift/origin-web-catalog/pull/501
![image](https://user-images.githubusercontent.com/11633780/31743536-74d5e69e-b429-11e7-92b7-9876807fda25.png)

IMO, either the background should be white or the entire area should be grey.

@beanh66

serenamarie125 commented:
Given the PatternFly guidelines I would make it grey.

amarie401 commented:
@serenamarie125 I agree

amarie401 commented:
With the grey background on ```.service-items``` & border on ```.pf-blank-slate```

<img width="1324" alt="screen shot 2017-10-18 at 6 11 15 pm" src="https://user-images.githubusercontent.com/20052391/31745221-d767bb6a-b42f-11e7-9c8a-4bee27f42d07.png">

With the grey background on ```.service-items``` &  border removed from ```.pf-blank-slate```

<img width="1319" alt="screen shot 2017-10-18 at 6 14 58 pm" src="https://user-images.githubusercontent.com/20052391/31745368-6e841476-b430-11e7-9698-e40f87db57f4.png">

serenamarie125 commented:
@amarie401 I'd expect that the toolbar would still have a white background, and that the area below the separator would be grey.  Does that make sense ?

dtaylor113 commented:
> You also need to include <pf-empty-state> for subcategories?

Hi @amarie401, just want to make sure this issue gets addressed.  You need to add a pf-empty-state [here](https://github.com/amarie401/origin-web-catalog/blob/48b80a3f3cca1bdc4c22e93e9ea693c0cf2b97c8/src/components/services-view/services-view.html#L78)  -thanks

amarie401 commented:
@rhamilto how could we go about making the empty state component span the full height. If we want to keep the toolbar and space around the empty state white?

rhamilto commented:
> @rhamilto how could we go about making the empty state component span the full height. If we want to keep the toolbar area white?

It's going to take some CSS hackin'.  We're doing a number of things to manipulate that box already (note the presence of "sass offerings" below the search bar impacts the way that white box renders).  I think the simplest approach is to not use the gray background as it alleviates the need to solve this problem and removes the inconsistency using the gray background introduces).

rhamilto commented:
This will turn off the background color and border for any pf-empty-state directives inside the service area.

```
.services-view .blank-slate-pf {
  background-color: transparent;
  border: none;
}
```

amarie401 commented:
I have updated the styling to use a transparent background as @rhamilto suggested and added the pfEmptyState component to the subcategory filters as well @dtaylor113

dtaylor113 commented:
LGTM, thanks!

spadgett commented:
@amarie401 I think you need to rebase to pick up the angular-patternfly version update. The "Clear Filters" link isn't working for me otherwise.

rhamilto commented on src/styles/services-view.less:
```diff
@@ -10,6 +10,14 @@ services-view,
 }
 
 .services-view {
+  .pf-empty-state {
```
Nit:  we alphabetize rules by selector, so this should go after .blank-slate-pf

amarie401 replied:
Okie dok will do!

amarie401 commented:
@spadgett I believe I had the updated version already on this current [branch](https://github.com/amarie401/origin-web-catalog/blob/emptystate-filter/package.json#L33).

spadgett commented:
Thanks @amarie401
